libs/cli/src/main/java/org/elasticsearch/cli/Terminal.java

```diff
@@ -339,4 +364,49 @@ public OutputStream getOutputStream() {
             return System.out;
         }
     }
+
+    /** A line based OutputStream wrapping this Terminal's println, not thread-safe! */
+    private class LineOutputStream extends OutputStream {
+        static final int DEFAULT_BUFFER_LENGTH = 1024;
+        static final int MAX_BUFFER_LENGTH = DEFAULT_BUFFER_LENGTH * 8;
+
+        private byte[] bytes = new byte[DEFAULT_BUFFER_LENGTH];
+        private int count = 0;
+
+        @Override
+        public void write(int b) {
+            if (b == 0) return;
+            if (b == '\n') {
+                flushLine(false);
+                return;
+            }
+            if (count == bytes.length) {
+                if (count >= MAX_BUFFER_LENGTH) {
+                    flushLine(false);
+                } else {
+                    bytes = Arrays.copyOf(bytes, 2 * bytes.length);
+                }
+            }
+            bytes[count++] = (byte) b;
+        }
+
+        private void flushLine(boolean skipEmpty) {
+            if (count > 0 && bytes[count - 1] == '\r') {
+                --count; // drop CR on windows as well
+            }
+            if (skipEmpty && count == 0) {
+                return;
+            }
+            println(count > 0 ? new String(bytes, 0, count, StandardCharsets.UTF_8) : "");
```

Comment:
Won't the charset be different on some systems eg windows? This seems system dependent.

Reply:
Currently it was used with `UTF_8` only:
https://github.com/elastic/elasticsearch/pull/106470/files#diff-c4ca52a921b22f0d3151383083ab03ae115d86fcb5c246b58b560ebd9f5fc12cL252-R251

But you're right, depending on the use case it could be another charset. I've extended the signature to path the charset 👍